### DISCUSSION TITLE
Processing several requests in the same task graph

### DISCUSSION THREAD

**Discussion by adpi2** [sbt/sbt#7246] [Ideas]
The BSP client often sends batches of requests all at once. For instance, Metals sends a `buildTarget/scalaMainClasses` a`buildTarget/scalaTestClasses` and a `buildTarget/jvmRunEnvironment` requests after each compilation. Processing each request sequentially takes time. We could save quite a lot of time if sbt could merge and process the requests in a single task graph.

The idea would be to introduce a new command that takes a list of tasks and run them sequentially in the same task graph, merging the common parts of the graph. This would only work if the tasks don't have side effects. For instance running `clean & compile & clean` in the same task graph (the second clean is a no-op) is not the same thing as running `clean; compile; clean`.

**Comment by eed3si9n** [comment]
There's `all` command.

**Reply by dragos** [reply]
This must be the understatement of the century :) Joke aside, I think even experienced developers have never seen it in action. Perhaps this could be considered as the default for batch mode (running `sbt compile test publish`) in 2.0. If users need sequential execution, this can easily be avoided via semicolons (`sbt "compile;test;publish"`)

There is still the thorny `clean`, but if my reading is correct, it is already [properly tagged](https://github.com/sbt/sbt/blob/e3b7870b2d98e26e38af82f5dd5e6b2b3f77b219/main/src/main/scala/sbt/Defaults.scala#L1658) and will not execute concurrently with any other task.
